File: src/domain/autoUpdate.ts
import {
  createAttentionNotification,
  type AttentionNotification,
} from './attentionNotifications'

export type AutoUpdateStatus =
  | 'idle'
  | 'checking'
  | 'up_to_date'
  | 'available'
  | 'downloading'
  | 'installing'
  | 'restart_required'
  | 'blocked'
  | 'error'

export type AutoUpdatePlatformPackage = {
  signature?: string | null
  url?: string | null
  size?: number | null
}

export type AutoUpdateManifest = {
  version?: string | null
  pubDate?: string | null
  notes?: string | null
  platforms?: Record<string, AutoUpdatePlatformPackage | undefined> | null
}

export type AutoUpdateCheckResult = {
  status: Extract<AutoUpdateStatus, 'up_to_date' | 'available' | 'blocked' | 'error'>
  currentVersion: string
  latestVersion: string | null
  checkedAt: string
  source: string
  releaseNotes: string | null
  packageSizeBytes: number | null
  signatureConfigured: boolean
  downloadUrl: string | null
  errorCode?: string | null
  errorMessage?: string | null
}

export type AutoUpdateState = {
  status: AutoUpdateStatus
  currentVersion: string
  latestVersion: string | null
  checkedAt: string | null
  source: string | null
  releaseNotes: string | null
  packageSizeBytes: number | null
  signatureConfigured: boolean
  downloadUrl: string | null
  downloadProgressPercent: number | null
  blockedReason: string | null
  message: string | null
  errorCode: string | null
}

export type AutoUpdateEvent =
  | { type: 'check_started'; checkedAt: string }
  | { type: 'check_succeeded'; result: AutoUpdateCheckResult }
  | { type: 'download_progress'; downloadedBytes: number; totalBytes: number | null }
  | { type: 'install_started' }
  | { type: 'install_blocked'; reason: string; message: string }
  | { type: 'install_finished'; restartRequired: boolean }
  | { type: 'failed'; errorCode: string; message: string }

export function createInitialAutoUpdateState(currentVersion: string): AutoUpdateState {
  return {
    status: 'idle',
    currentVersion,
    latestVersion: null,
    checkedAt: null,
    source: null,
    releaseNotes: null,
    packageSizeBytes: null,
    signatureConfigured: false,
    downloadUrl: null,
    downloadProgressPercent: null,
    blockedReason: null,
    message: null,
    errorCode: null,
  }
}

export function evaluateUpdateManifest(input: {
  currentVersion: string
  currentPlatform: string
  manifest: AutoUpdateManifest
  checkedAt: string
  source: string
}): AutoUpdateCheckResult {
  const latestVersion = input.manifest.version?.trim() || null
  const platformPackage = input.manifest.platforms?.[input.currentPlatform] ?? null

  if (!latestVersion) {
    return createBlockedResult(input, null, platformPackage, 'missing_version', '更新清单缺少版本号。')
  }

  if (!platformPackage) {
    return createBlockedResult(input, latestVersion, null, 'platform_not_supported', '更新清单缺少当前 macOS 架构包。')
  }

  const signatureConfigured = Boolean(platformPackage.signature?.trim())
  if (!signatureConfigured) {
    return createBlockedResult(input, latestVersion, platformPackage, 'missing_signature', '更新包缺少签名，已阻止安装。')
  }

  if (!platformPackage.url?.trim()) {
    return createBlockedResult(input, latestVersion, platformPackage, 'missing_download_url', '更新包缺少下载地址。')
  }

  if (compareUpdateVersions(input.currentVersion, latestVersion) >= 0) {
    return {
      status: 'up_to_date',
      currentVersion: input.currentVersion,
      latestVersion,
      checkedAt: input.checkedAt,
      source: input.source,
      releaseNotes: input.manifest.notes ?? null,
      packageSizeBytes: normalizePackageSize(platformPackage.size),
      signatureConfigured,
      downloadUrl: platformPackage.url,
    }
  }

  return {
    status: 'available',
    currentVersion: input.currentVersion,
    latestVersion,
    checkedAt: input.checkedAt,
    source: input.source,
    releaseNotes: input.manifest.notes ?? null,
    packageSizeBytes: normalizePackageSize(platformPackage.size),
    signatureConfigured,
    downloadUrl: platformPackage.url,
  }
}

export function compareUpdateVersions(left: string, right: string): number {
  const leftParts = parseVersionParts(left)
  const rightParts = parseVersionParts(right)
  const length = Math.max(leftParts.length, rightParts.length)

  for (let index = 0; index < length; index += 1) {
    const leftPart = leftParts[index] ?? 0
    const rightPart = rightParts[index] ?? 0
    if (leftPart > rightPart) {
      return 1
    }
    if (leftPart < rightPart) {
      return -1
    }
  }

  return 0
}

export function reduceAutoUpdateEvent(state: AutoUpdateState, event: AutoUpdateEvent): AutoUpdateState {
  switch (event.type) {
    case 'check_started':
      return {
        ...state,
        status: 'checking',
        checkedAt: event.checkedAt,
        message: '正在检查更新。',
        errorCode: null,
      }
    case 'check_succeeded':
      return {
        ...state,
        status: event.result.status,
        latestVersion: event.result.latestVersion,
        checkedAt: event.result.checkedAt,
        source: event.result.source,
        releaseNotes: event.result.releaseNotes,
        packageSizeBytes: event.result.packageSizeBytes,
        signatureConfigured: event.result.signatureConfigured,
        downloadUrl: event.result.downloadUrl,
        blockedReason: event.result.status === 'blocked' ? event.result.errorCode ?? null : null,
        message: event.result.errorMessage ?? null,
        errorCode: event.result.errorCode ?? null,
      }
    case 'download_progress':
      return {
        ...state,
        status: 'downloading',
        downloadProgressPercent:
          event.totalBytes && event.totalBytes > 0
            ? Math.min(100, Math.max(0, Math.round((event.downloadedBytes / event.totalBytes) * 100)))
            : null,
        message: '正在下载更新。',
      }
    case 'install_started':
      return {
        ...state,
        status: 'installing',
        message: '正在安装更新。',
      }
    case 'install_blocked':
      return {
        ...state,
        status: 'blocked',
        blockedReason: event.reason,
        message: event.message,
        errorCode: event.reason,
      }
    case 'install_finished':
      return {
        ...state,
        status: event.restartRequired ? 'restart_required' : 'up_to_date',
        message: event.restartRequired ? '安装完成，需要重启 Aiotto。' : '安装完成。',
      }
    case 'failed':
      return {
        ...state,
        status: 'error',
        errorCode: event.errorCode,
        message: event.message,
      }
  }
}

export function buildUpdateNotifications(
  results: AutoUpdateCheckResult[],
  createdAt: string,
): AttentionNotification[] {
  return results.flatMap((result) => {
    if (result.status === 'available') {
      return [
        createAttentionNotification({
          id: `update-available-${result.latestVersion ?? 'unknown'}`,
          type: 'update_available',
          priority: 'medium',
          title: 'Aiotto 有新版本',
          body: `${result.currentVersion} -> ${result.latestVersion ?? 'Unknown'}，可在版本与更新中查看说明并安装。`,
          createdAt,
          targetPage: 'settings',
          targetId: 'settings:update',
          targetLabel: '版本与更新',
          system: true,
        }),
      ]
    }

    if (result.status === 'blocked') {
      return [
        createAttentionNotification({
          id: `update-blocked-${result.errorCode ?? 'unknown'}`,
          type: 'update_blocked',
          priority: 'high',
          title: '自动更新已阻止',
          body: result.errorMessage ?? '更新安装前检查未通过。',
          createdAt,
          targetPage: 'settings',
          targetId: 'settings:update',
          targetLabel: '版本与更新',
          system: true,
        }),
      ]
    }

    if (result.status === 'error') {
      return [
        createAttentionNotification({
          id: `update-failed-${result.errorCode ?? 'unknown'}`,
          type: 'update_failed',
          priority: 'high',
          title: '自动更新失败',
          body: result.errorMessage ?? '更新检查失败。',
          createdAt,
          targetPage: 'settings',
          targetId: 'settings:update',
          targetLabel: '版本与更新',
          system: true,
        }),
      ]
    }

    return []
  })
}

function createBlockedResult(
  input: {
    currentVersion: string
    manifest: AutoUpdateManifest
    checkedAt: string
    source: string
  },
  latestVersion: string | null,
  platformPackage: AutoUpdatePlatformPackage | null,
  errorCode: string,
  errorMessage: string,
): AutoUpdateCheckResult {
  return {
    status: 'blocked',
    currentVersion: input.currentVersion,
    latestVersion,
    checkedAt: input.checkedAt,
    source: input.source,
    releaseNotes: input.manifest.notes ?? null,
    packageSizeBytes: normalizePackageSize(platformPackage?.size),
    signatureConfigured: Boolean(platformPackage?.signature?.trim()),
    downloadUrl: platformPackage?.url?.trim() || null,
    errorCode,
    errorMessage,
  }
}

function normalizePackageSize(size: number | null | undefined): number | null {
  return Number.isFinite(size) && Number(size) > 0 ? Number(size) : null
}

function parseVersionParts(version: string): number[] {
  return version
    .replace(/^[^\d]*/, '')
    .split(/[.-]/)
    .map((part) => Number.parseInt(part, 10))
    .map((part) => (Number.isFinite(part) ? part : 0))
}

File: apps/desktop/src-tauri/src/auto_update.rs
use serde::{Deserialize, Serialize};
use std::{
    collections::BTreeMap,
    fs,
    path::{Path, PathBuf},
};
use tauri::{AppHandle, Emitter};
use tauri_plugin_updater::UpdaterExt;

const TAURI_UPDATE_SOURCE: &str =
    "https://github.com/ShawnZhuge/aiotto/releases/latest/download/latest.json";
const TAURI_UPDATE_DOWNLOAD_EVENT: &str = "auto-update-download-progress";

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct AutoUpdateManifest {
    pub version: String,
    #[serde(default, alias = "pub_date")]
    pub pub_date: Option<String>,
    #[serde(default)]
    pub notes: Option<String>,
    #[serde(default)]
    pub platforms: BTreeMap<String, AutoUpdatePlatformPackage>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct AutoUpdatePlatformPackage {
    #[serde(default)]
    pub signature: Option<String>,
    #[serde(default)]
    pub url: Option<String>,
    #[serde(default)]
    pub size: Option<u64>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct AutoUpdateCheckResult {
    pub status: String,
    pub current_version: String,
    pub latest_version: Option<String>,
    pub checked_at: Option<String>,
    pub source: String,
    pub release_notes: Option<String>,
    pub package_size_bytes: Option<u64>,
    pub signature_configured: bool,
    pub download_url: Option<String>,
    pub error_code: Option<String>,
    pub error_message: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct InstallPreflightResult {
    pub status: String,
    pub reason: Option<String>,
    pub message: String,
    pub bundle_path: String,
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CheckAutoUpdateManifestRequest {
    pub latest_json_path: String,
    pub current_version: String,
    pub platform: String,
}

#[derive(Debug, Clone, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct PrepareAutoUpdateInstallRequest {
    pub package_url: String,
    pub version: String,
    pub signature: String,
    pub expected_signature: String,
    pub staging_dir: String,
    pub bundle_path: String,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct PrepareAutoUpdateInstallResult {
    pub status: String,
    pub reason: Option<String>,
    pub message: String,
    pub version: String,
    pub staged_package_path: String,
    pub receipt_path: String,
    pub restart_required: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct AutoUpdateDownloadProgress {
    pub downloaded: u64,
    pub total: Option<u64>,
}

pub fn check_auto_update_manifest_in(
    latest_json_path: &Path,
    current_version: &str,
    platform: &str,
) -> Result<AutoUpdateCheckResult, String> {
    let raw = fs::read_to_string(latest_json_path).map_err(|error| error.to_string())?;
    let manifest = serde_json::from_str::<AutoUpdateManifest>(&raw)
        .map_err(|error| format!("latest.json 解析失败：{error}"))?;
    Ok(evaluate_manifest(
        manifest,
        current_version,
        platform,
        latest_json_path.to_string_lossy().to_string(),
    ))
}

pub fn check_install_preflight_for_bundle_path(bundle_path: &Path) -> InstallPreflightResult {
    let bundle_path_label = bundle_path.to_string_lossy().to_string();
    let lower_path = bundle_path_label.to_ascii_lowercase();

    if bundle_path_label.starts_with("/Volumes/") {
        return blocked_preflight(
            bundle_path_label,
            "running_from_dmg",
            "当前从 DMG 运行，请拖入 Applications 后重新打开。",
        );
    }

    if lower_path.contains("apptranslocation") {
        return blocked_preflight(
            bundle_path_label,
            "app_translocation",
            "当前命中 App Translocation，请把 Aiotto 移入 Applications 后重新打开。",
        );
    }

    if !bundle_path_label.contains(".app/") && !bundle_path_label.ends_with(".app") {
        return blocked_preflight(
            bundle_path_label,
            "not_app_bundle",
            "未检测到 .app bundle 路径，不能执行自动替换安装。",
        );
    }

    InstallPreflightResult {
        status: "ready".to_string(),
        reason: None,
        message: "安装位置可用于自动更新。".to_string(),
        bundle_path: bundle_path_label,
    }
}

#[tauri::command]
pub async fn check_auto_update(app: AppHandle) -> Result<AutoUpdateCheckResult, String> {
    let current_version = app.package_info().version.to_string();
    let checked_at = chrono::Utc::now().to_rfc3339();
    let updater = app
        .updater_builder()
        .build()
        .map_err(|error| format!("初始化更新器失败：{error}"))?;

    match updater
        .check()
        .await
        .map_err(|error| format!("检查更新失败：{error}"))?
    {
        Some(update) => Ok(AutoUpdateCheckResult {
            status: "available".to_string(),
            current_version,
            latest_version: Some(update.version.clone()),
            checked_at: update
                .date
                .as_ref()
                .map(|date| date.to_string())
                .or(Some(checked_at)),
            source: TAURI_UPDATE_SOURCE.to_string(),
            release_notes: update.body.clone(),
            package_size_bytes: None,
            signature_configured: !update.signature.trim().is_empty(),
            download_url: Some(update.download_url.to_string()),
            error_code: None,
            error_message: None,
        }),
        None => Ok(AutoUpdateCheckResult {
            status: "up_to_date".to_string(),
            current_version: current_version.clone(),
            latest_version: Some(current_version),
            checked_at: Some(checked_at),
            source: TAURI_UPDATE_SOURCE.to_string(),
            release_notes: None,
            package_size_bytes: None,
            signature_configured: true,
            download_url: None,
            error_code: None,
            error_message: None,
        }),
    }
}

#[tauri::command]
pub fn check_auto_update_manifest(
    request: CheckAutoUpdateManifestRequest,
) -> Result<AutoUpdateCheckResult, String> {
    check_auto_update_manifest_in(
        Path::new(&request.latest_json_path),
        &request.current_version,
        &request.platform,
    )
}

#[tauri::command]
pub async fn install_auto_update_and_restart(
    app: AppHandle,
) -> Result<PrepareAutoUpdateInstallResult, String> {
    let updater = app
        .updater_builder()
        .build()
        .map_err(|error| format!("初始化更新器失败：{error}"))?;
    let Some(update) = updater
        .check()
        .await
        .map_err(|error| format!("检查更新失败：{error}"))?
    else {
        return Ok(PrepareAutoUpdateInstallResult {
            status: "up_to_date".to_string(),
            reason: None,
            message: "当前已是最新版本。".to_string(),
            version: app.package_info().version.to_string(),
            staged_package_path: String::new(),
            receipt_path: String::new(),
            restart_required: false,
        });
    };

    let version = update.version.clone();
    log::info!("开始下载 Aiotto 更新: {version}");
    let progress_app = app.clone();
    let mut downloaded: u64 = 0;
    let bytes = update
        .download(
            move |chunk_len, content_len| {
                downloaded = downloaded.saturating_add(chunk_len as u64);
                let _ = progress_app.emit(
                    TAURI_UPDATE_DOWNLOAD_EVENT,
                    AutoUpdateDownloadProgress {
                        downloaded,
                        total: content_len,
                    },
                );
            },
            || {},
        )
        .await
        .map_err(|error| format!("下载更新失败：{error}"))?;

    update
        .install(bytes)
        .map_err(|error| format!("安装更新失败：{error}"))?;

    log::info!("Aiotto 更新已安装，正在重启应用: {version}");
    app.restart();
}

#[tauri::command]
pub fn check_auto_update_install_preflight(
    bundle_path: Option<String>,
) -> Result<InstallPreflightResult, String> {
    let resolved = bundle_path
        .map(PathBuf::from)
        .or_else(|| std::env::current_exe().ok())
        .ok_or_else(|| "无法解析当前 Aiotto bundle 路径。".to_string())?;
    Ok(check_install_preflight_for_bundle_path(&resolved))
}

pub fn prepare_auto_update_install_in(
    request: PrepareAutoUpdateInstallRequest,
) -> Result<PrepareAutoUpdateInstallResult, String> {
    if request.signature.trim().is_empty() || request.expected_signature.trim().is_empty() {
        return Ok(blocked_prepare_result(
            request.version,
            "missing_signature",
            "更新包缺少签名，已阻止安装。",
            request.staging_dir,
        ));
    }

    if request.signature.trim() != request.expected_signature.trim() {
        return Ok(blocked_prepare_result(
            request.version,
            "signature_mismatch",
            "更新包签名不匹配，已阻止安装。",
            request.staging_dir,
        ));
    }

    let preflight = check_install_preflight_for_bundle_path(Path::new(&request.bundle_path));
    if preflight.status == "blocked" {
        return Ok(blocked_prepare_result(
            request.version,
            preflight
                .reason
                .as_deref()
                .unwrap_or("install_preflight_blocked"),
            &preflight.message,
            request.staging_dir,
        ));
    }

    let package_path = resolve_sandbox_package_path(&request.package_url)?;
    if !package_path.exists() {
        return Ok(blocked_prepare_result(
            request.version,
            "package_not_found",
            "未找到沙盒更新包，已阻止安装。",
            request.staging_dir,
        ));
    }

    let staging_dir = PathBuf::from(&request.staging_dir);
    fs::create_dir_all(&staging_dir).map_err(|error| error.to_string())?;
    let file_name = package_path
        .file_name()
        .and_then(|value| value.to_str())
        .unwrap_or("Aiotto-update-package");
    let staged_package_path = staging_dir.join(file_name);
    fs::copy(&package_path, &staged_package_path).map_err(|error| error.to_string())?;

    let receipt_path = staging_dir.join("install-receipt.json");
    let receipt = serde_json::json!({
        "version": request.version,
        "stagedPackagePath": staged_package_path,
        "bundlePath": request.bundle_path,
        "restartRequired": true,
        "mode": "sandbox_prepare_only"
    });
    fs::write(
        &receipt_path,
        serde_json::to_string_pretty(&receipt).map_err(|error| error.to_string())?,
    )
    .map_err(|error| error.to_string())?;

    Ok(PrepareAutoUpdateInstallResult {
        status: "restart_required".to_string(),
        reason: None,
        message: "沙盒更新包已准备，重启后可执行安装验证。".to_string(),
        version: request.version,
        staged_package_path: staged_package_path.to_string_lossy().to_string(),
        receipt_path: receipt_path.to_string_lossy().to_string(),
        restart_required: true,
    })
}

#[tauri::command]
pub fn prepare_auto_update_install(
    request: PrepareAutoUpdateInstallRequest,
) -> Result<PrepareAutoUpdateInstallResult, String> {
    prepare_auto_update_install_in(request)
}

fn evaluate_manifest(
    manifest: AutoUpdateManifest,
    current_version: &str,
    platform: &str,
    source: String,
) -> AutoUpdateCheckResult {
    let package = manifest.platforms.get(platform);
    if package.is_none() {
        return blocked_result(
            current_version,
            Some(manifest.version),
            manifest.pub_date,
            source,
            manifest.notes,
            None,
            false,
            None,
            "platform_not_supported",
            "更新清单缺少当前 macOS 架构包。",
        );
    }

    let package = package.expect("checked above");
    let signature = package.signature.as_deref().unwrap_or("").trim();
    if signature.is_empty() {
        return blocked_result(
            current_version,
            Some(manifest.version),
            manifest.pub_date,
            source,
            manifest.notes,
            package.size,
            false,
            package.url.clone(),
            "missing_signature",
            "更新包缺少签名，已阻止安装。",
        );
    }

    let url = package.url.as_deref().unwrap_or("").trim();
    if url.is_empty() {
        return blocked_result(
            current_version,
            Some(manifest.version),
            manifest.pub_date,
            source,
            manifest.notes,
            package.size,
            true,
            None,
            "missing_download_url",
            "更新包缺少下载地址。",
        );
    }

    let status = if compare_versions(&manifest.version, current_version) > 0 {
        "available"
    } else {
        "up_to_date"
    };

    AutoUpdateCheckResult {
        status: status.to_string(),
        current_version: current_version.to_string(),
        latest_version: Some(manifest.version),
        checked_at: manifest.pub_date,
        source,
        release_notes: manifest.notes,
        package_size_bytes: package.size,
        signature_configured: true,
        download_url: Some(url.to_string()),
        error_code: None,
        error_message: None,
    }
}

#[allow(clippy::too_many_arguments)]
fn blocked_result(
    current_version: &str,
    latest_version: Option<String>,
    checked_at: Option<String>,
    source: String,
    release_notes: Option<String>,
    package_size_bytes: Option<u64>,
    signature_configured: bool,
    download_url: Option<String>,
    error_code: &str,
    error_message: &str,
) -> AutoUpdateCheckResult {
    AutoUpdateCheckResult {
        status: "blocked".to_string(),
        current_version: current_version.to_string(),
        latest_version,
        checked_at,
        source,
        release_notes,
        package_size_bytes,
        signature_configured,
        download_url,
        error_code: Some(error_code.to_string()),
        error_message: Some(error_message.to_string()),
    }
}

fn blocked_preflight(bundle_path: String, reason: &str, message: &str) -> InstallPreflightResult {
    InstallPreflightResult {
        status: "blocked".to_string(),
        reason: Some(reason.to_string()),
        message: message.to_string(),
        bundle_path,
    }
}

fn blocked_prepare_result(
    version: String,
    reason: &str,
    message: &str,
    staging_dir: String,
) -> PrepareAutoUpdateInstallResult {
    PrepareAutoUpdateInstallResult {
        status: "blocked".to_string(),
        reason: Some(reason.to_string()),
        message: message.to_string(),
        version,
        staged_package_path: String::new(),
        receipt_path: PathBuf::from(staging_dir)
            .join("install-receipt.json")
            .to_string_lossy()
            .to_string(),
        restart_required: false,
    }
}

fn resolve_sandbox_package_path(package_url: &str) -> Result<PathBuf, String> {
    if let Some(path) = package_url.strip_prefix("file://") {
        return Ok(PathBuf::from(path));
    }

    if package_url.starts_with("http://") || package_url.starts_with("https://") {
        return Err(
            "沙盒安装只允许 file:// 或本地路径，真实下载留给发布包 updater 验证。".to_string(),
        );
    }

    Ok(PathBuf::from(package_url))
}

fn compare_versions(left: &str, right: &str) -> i8 {
    let left_parts = version_parts(left);
    let right_parts = version_parts(right);
    let max_len = left_parts.len().max(right_parts.len());

    for index in 0..max_len {
        let left_part = *left_parts.get(index).unwrap_or(&0);
        let right_part = *right_parts.get(index).unwrap_or(&0);
        if left_part > right_part {
            return 1;
        }
        if left_part < right_part {
            return -1;
        }
    }

    0
}

fn version_parts(version: &str) -> Vec<u64> {
    version
        .trim_start_matches(|character: char| !character.is_ascii_digit())
        .split(['.', '-'])
        .map(|part| part.parse::<u64>().unwrap_or(0))
        .collect()
}
